Document/0x05j-Testing-Resiliency-Against-Reverse-Engineering.md

```diff
@@ -675,17 +677,125 @@ private void crcTest() throws IOException {
  }
 }
 ```
+##### Sample Implementation - Storage
+
+When providing integrity on the file-storage itself. You can either create an HMAC over a given key-value pair as for the Android `SharedPreferences` or you can create an HMAC over a complete file provided by the filesystem.
+When using an HMAC, you can either use a bouncy castle implementation to HMAC the given content or the AndroidKeyStore and then verify the HMAC later on. There are a few steps to take care off.
+In case of the need for encryptionl. Please make sure that you encrypt and then HMAC as described in [2].
+
+When generating an HMAC with BC:
+
+1. Make sure BounceyCastle or SpongeyCastle are registered as a security provider.
+2. Initialize the HMAC with a key, which can be stored in a keystore.
+3. Get the bytearray of the content that needs an HMAC.
+4. Call `doFinal` on the HMAC with the bytecode.
+5. Append the HMAC to the bytearray of step 3.
+6. Store the result of step 5.
+
+When verifying the HMAC with BC:
+
+1. Make sure BounceyCastle or SpongeyCastle are registered as a security provider.
+2. Extract the message and the hmacbytes as separate arrays.
+3. repeate step 1-4 of generating an hmac on the message.
+4. now compare the extracted hamcbytes to the result of step 3.
+
+When generating the HMAC based on the Android keystore, then it is best to only do this for Android 6 and higher. In that case you generate the key for hmacking as described in [3].
+A convinient HMAC implementation without the `AndroidKeyStore` can be found below:
+
+```java
+public enum HMACWrapper {
+    HMAC_512("HMac-SHA512"),
```

Comment:
FYI: turns out capitalization and presence/absence of the dash ("-") does not matter, at least on the Nougat emulator.